pkg/scalers/azure_log_analytics_scaler.go

```diff
@@ -0,0 +1,492 @@
+		logAnalyticsLog.Error(err, "Error processing Log Analytics query")
+		return false, err
+	}
+
+	return s.cache.metricValue > 0, nil
+}
+
+func (s *azureLogAnalyticsScaler) GetMetricSpecForScaling() []v2beta2.MetricSpec {
+	err := s.updateCache()
+
+	if err != nil {
+		logAnalyticsLog.Error(err, "Error processing Log Analytics query")
+		return nil
+	}
+
+	externalMetric := &v2beta2.ExternalMetricSource{
+		Metric: v2beta2.MetricIdentifier{
+			Name: fmt.Sprintf("%s-%s", "azure-log-analytics", s.metadata.workspaceID),
+		},
+		Target: v2beta2.MetricTarget{
+			Type:         v2beta2.AverageValueMetricType,
+			AverageValue: resource.NewQuantity(s.cache.metricThreshold, resource.DecimalSI),
+		},
+	}
+	metricSpec := v2beta2.MetricSpec{External: externalMetric, Type: externalMetricType}
+	return []v2beta2.MetricSpec{metricSpec}
+}
+
+//GetMetrics returns value for a supported metric and an error if there is a problem getting the metric
+func (s *azureLogAnalyticsScaler) GetMetrics(ctx context.Context, metricName string, metricSelector labels.Selector) ([]external_metrics.ExternalMetricValue, error) {
+	receivedMetric, err := s.getMetricData()
+
+	if err != nil {
+		logAnalyticsLog.Error(err, "Failed to get metrics.")
+		return []external_metrics.ExternalMetricValue{}, err
+	}
+
+	metric := external_metrics.ExternalMetricValue{
+		MetricName: metricName,
+		Value:      *resource.NewQuantity(receivedMetric.value, resource.DecimalSI),
+		Timestamp:  metav1.Now(),
+	}
+
+	return append([]external_metrics.ExternalMetricValue{}, metric), nil
+}
+
+func (s *azureLogAnalyticsScaler) Close() error {
+	return nil
+}
+
+func (s *azureLogAnalyticsScaler) updateCache() error {
+	if s.cache.metricValue < 0 {
+		receivedMetric, err := s.getMetricData()
+
+		if err != nil {
+			logAnalyticsLog.Error(err, "Error processing Log Analytics query")
+			return err
+		}
+
+		s.cache.metricValue = receivedMetric.value
+
+		if receivedMetric.threshold > 0 {
+			s.cache.metricThreshold = receivedMetric.threshold
+		} else {
+			s.cache.metricThreshold = s.metadata.threshold
+		}
+	}
+
+	return nil
+}
+
+func (s *azureLogAnalyticsScaler) getMetricData() (metricsData, error) {
+	tokenInfo, err := s.getAccessToken()
+	if err != nil {
+		return metricsData{}, err
+	}
+
+	metricsInfo, err := s.executeQuery(s.metadata.query, tokenInfo)
+	if err != nil {
+		return metricsData{}, err
+	}
+
+	logAnalyticsLog.Info(fmt.Sprintf("Providing metric value: %d", metricsInfo.value))
+	return metricsInfo, nil
+}
+
+func (s *azureLogAnalyticsScaler) getAccessToken() (tokenData, error) {
+	//if there is no token yet or it will be expired in less, that 30 secs
+	currentTimeSec := time.Now().Unix()
+	if currentTimeSec+30 > s.cache.accessToken.ExpiresOn {
+		tokenInfo, err := s.refreshAccessToken()
+		if err != nil {
+			return tokenData{}, err
+		}
+
+		s.cache.accessToken = tokenInfo
+		return tokenInfo, nil
+	}
+
+	logAnalyticsLog.Info(fmt.Sprintf("Reuse existing access token..."))
```

Comment:
No need to call fmt.Sprintf() inside.

Reply:
Thanks for pointing. I have removed this entry in last commit, as it is not so valuable in this case.